File: examples/s3/s3_client.go
package main

import (
	"bytes"
	"fmt"
	"io"
	"log"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/credentials"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/s3"
)

const (
	endpoint  = "http://localhost:9000"
	accessKey = "minioadmin"
	secretKey = "minioadmin"
	region    = "us-east-1"
	bucket    = "test-bucket"
)

func main() {
	// Create S3 session
	sess, err := session.NewSession(&aws.Config{
		Endpoint:         aws.String(endpoint),
		Region:           aws.String(region),
		Credentials:      credentials.NewStaticCredentials(accessKey, secretKey, ""),
		S3ForcePathStyle: aws.Bool(true),
	})
	if err != nil {
		log.Fatal("Failed to create session:", err)
	}

	svc := s3.New(sess)

	fmt.Println("S3-Compatible Storage Go Client Example")
	fmt.Println("========================================")

	// 1. Create bucket
	fmt.Printf("1. Creating bucket '%s'...\n", bucket)
	_, err = svc.CreateBucket(&s3.CreateBucketInput{
		Bucket: aws.String(bucket),
	})
	if err != nil {
		log.Printf("Failed to create bucket: %v", err)
	} else {
		fmt.Println("   ✓ Bucket created")
	}

	// 2. List buckets
	fmt.Println("\n2. Listing all buckets...")
	result, err := svc.ListBuckets(nil)
	if err != nil {
		log.Printf("Failed to list buckets: %v", err)
	} else {
		for _, b := range result.Buckets {
			fmt.Printf("   - %s\n", aws.StringValue(b.Name))
		}
	}

	// 3. Upload file
	fmt.Println("\n3. Uploading file...")
	content := []byte("Hello from Go S3 client!")
	_, err = svc.PutObject(&s3.PutObjectInput{
		Bucket:      aws.String(bucket),
		Key:         aws.String("test-go.txt"),
		Body:        bytes.NewReader(content),
		ContentType: aws.String("text/plain"),
	})
	if err != nil {
		log.Printf("Failed to upload file: %v", err)
	} else {
		fmt.Println("   ✓ File uploaded")
	}

	// 4. List objects
	fmt.Printf("\n4. Listing objects in '%s'...\n", bucket)
	listResult, err := svc.ListObjectsV2(&s3.ListObjectsV2Input{
		Bucket: aws.String(bucket),
	})
	if err != nil {
		log.Printf("Failed to list objects: %v", err)
	} else {
		for _, obj := range listResult.Contents {
			fmt.Printf("   - %s (Size: %d bytes)\n",
				aws.StringValue(obj.Key),
				aws.Int64Value(obj.Size))
		}
	}

	// 5. Download file
	fmt.Println("\n5. Downloading file...")
	getResult, err := svc.GetObject(&s3.GetObjectInput{
		Bucket: aws.String(bucket),
		Key:    aws.String("test-go.txt"),
	})
	if err != nil {
		log.Printf("Failed to download file: %v", err)
	} else {
		defer getResult.Body.Close()
		downloadedContent, _ := io.ReadAll(getResult.Body)
		fmt.Printf("   Downloaded content: %s\n", string(downloadedContent))
	}

	// 6. Get object metadata
	fmt.Println("\n6. Getting object metadata...")
	headResult, err := svc.HeadObject(&s3.HeadObjectInput{
		Bucket: aws.String(bucket),
		Key:    aws.String("test-go.txt"),
	})
	if err != nil {
		log.Printf("Failed to get metadata: %v", err)
	} else {
		fmt.Printf("   - ETag: %s\n", aws.StringValue(headResult.ETag))
		fmt.Printf("   - Content-Type: %s\n", aws.StringValue(headResult.ContentType))
		fmt.Printf("   - Content-Length: %d\n", aws.Int64Value(headResult.ContentLength))
	}

	// 7. Delete object
	fmt.Println("\n7. Deleting object...")
	_, err = svc.DeleteObject(&s3.DeleteObjectInput{
		Bucket: aws.String(bucket),
		Key:    aws.String("test-go.txt"),
	})
	if err != nil {
		log.Printf("Failed to delete object: %v", err)
	} else {
		fmt.Println("   ✓ Object deleted")
	}

	// 8. Delete bucket
	fmt.Printf("\n8. Deleting bucket '%s'...\n", bucket)
	_, err = svc.DeleteBucket(&s3.DeleteBucketInput{
		Bucket: aws.String(bucket),
	})
	if err != nil {
		log.Printf("Failed to delete bucket: %v", err)
	} else {
		fmt.Println("   ✓ Bucket deleted")
	}

	fmt.Println("\n========================================")
	fmt.Println("✓ All tests completed!")
}

File: examples/s3/s3_client.py
#!/usr/bin/env python3
"""
S3-compatible storage client example
Requires: pip install boto3
"""

import boto3
from botocore.client import Config

# Configuration
S3_ENDPOINT = 'http://localhost:9000'
ACCESS_KEY = 'minioadmin'
SECRET_KEY = 'minioadmin'
BUCKET_NAME = 'test-bucket'

def main():
    # Create S3 client
    s3_client = boto3.client(
        's3',
        endpoint_url=S3_ENDPOINT,
        aws_access_key_id=ACCESS_KEY,
        aws_secret_access_key=SECRET_KEY,
        config=Config(signature_version='s3v4'),
        region_name='us-east-1'
    )
    
    print("S3-Compatible Storage Client Example")
    print("=" * 40)
    
    try:
        # 1. Create bucket
        print(f"1. Creating bucket '{BUCKET_NAME}'...")
        s3_client.create_bucket(Bucket=BUCKET_NAME)
        print("   ✓ Bucket created")
        
        # 2. List buckets
        print("\n2. Listing all buckets...")
        response = s3_client.list_buckets()
        for bucket in response['Buckets']:
            print(f"   - {bucket['Name']}")
        
        # 3. Upload file
        print("\n3. Uploading file...")
        test_content = b"Hello from Python S3 client!"
        s3_client.put_object(
            Bucket=BUCKET_NAME,
            Key='test-python.txt',
            Body=test_content,
            ContentType='text/plain'
        )
        print("   ✓ File uploaded")
        
        # 4. List objects in bucket
        print(f"\n4. Listing objects in '{BUCKET_NAME}'...")
        response = s3_client.list_objects_v2(Bucket=BUCKET_NAME)
        if 'Contents' in response:
            for obj in response['Contents']:
                print(f"   - {obj['Key']} (Size: {obj['Size']} bytes)")
        
        # 5. Download file
        print("\n5. Downloading file...")
        response = s3_client.get_object(Bucket=BUCKET_NAME, Key='test-python.txt')
        downloaded_content = response['Body'].read()
        print(f"   Downloaded content: {downloaded_content.decode()}")
        
        # 6. Get object metadata
        print("\n6. Getting object metadata...")
        response = s3_client.head_object(Bucket=BUCKET_NAME, Key='test-python.txt')
        print(f"   - ETag: {response['ETag']}")
        print(f"   - Content-Type: {response['ContentType']}")
        print(f"   - Content-Length: {response['ContentLength']}")
        
        # 7. Delete object
        print("\n7. Deleting object...")
        s3_client.delete_object(Bucket=BUCKET_NAME, Key='test-python.txt')
        print("   ✓ Object deleted")
        
        # 8. Delete bucket
        print(f"\n8. Deleting bucket '{BUCKET_NAME}'...")
        s3_client.delete_bucket(Bucket=BUCKET_NAME)
        print("   ✓ Bucket deleted")
        
    except Exception as e:
        print(f"\n❌ Error: {e}")
        return 1
    
    print("\n" + "=" * 40)
    print("✓ All tests completed successfully!")
    return 0

if __name__ == '__main__':
    exit(main())